src/llama.cpp

```diff
@@ -18436,282 +18517,185 @@ static void llama_model_quantize_internal(const std::string & fname_inp, const s
-                return 1;
+    // bundle lora_a and lora_b into pairs
+    std::map<std::string, lora_weight> ab_map;
+    auto str_endswith = [](const std::string & str, const std::string & suffix) {
+        return str.size() >= suffix.size() && str.compare(str.size()-suffix.size(), suffix.size(), suffix) == 0;
+    };
+    for (ggml_tensor * cur = ggml_get_first_tensor(ctx); cur; cur = ggml_get_next_tensor(ctx, cur)) {
+        std::string name(cur->name);
+        if (str_endswith(name, ".lora_a")) {
+            replace_all(name, ".lora_a", "");
+            if (ab_map.find(name) == ab_map.end()) {
+                ab_map[name] = lora_weight(cur, nullptr);
+            } else {
+                ab_map[name].a = cur;
+            }
+        } else if (str_endswith(name, ".lora_b")) {
+            replace_all(name, ".lora_b", "");
+            if (ab_map.find(name) == ab_map.end()) {
+                ab_map[name] = lora_weight(nullptr, cur);
+            } else {
+                ab_map[name].b = cur;
             }
-            base_t = ggml_dup_tensor(lora_ctx, ml->get_tensor_meta(base_name.c_str()));
         } else {
-            base_t = ggml_dup_tensor(lora_ctx, model_t);
-        }
-        ggml_set_name(base_t, base_name.c_str());
-
-        // allocate in backend buffer
-        ggml_backend_buffer_t lora_buf = ggml_backend_alloc_ctx_tensors_from_buft(lora_ctx, ggml_backend_cpu_buffer_type());
-        if (lora_buf == nullptr) {
-            LLAMA_LOG_ERROR("%s: error: failed to allocate lora tensors\n", __func__);
-            return 1;
+            LLAMA_LOG_WARN("%s: discard tensor '%s'\n", __func__, cur->name);
         }
+    }
 
-        // load tensor data
-        auto load_tensor = [&read_buf, &fin](const tensor_meta & tensor_meta, ggml_tensor * tensor) {
-            read_buf.resize(ggml_nbytes(tensor));
-            fin.seek(tensor_meta.offset, SEEK_SET);
-            fin.read_raw(read_buf.data(), ggml_nbytes(tensor));
-            ggml_backend_tensor_set(tensor, read_buf.data(), 0, read_buf.size());
-        };
-        load_tensor(metaA, loraA);
-        load_tensor(metaB, loraB);
-
-        // load base model tensor data
-        if (ml) {
-            ml->load_data_for(base_t);
+    // add tensors
+    for (auto & it : ab_map) {
+        std::string name = it.first;
+        const char * cname = name.c_str();
+        lora_weight & w = it.second;
+        GGML_ASSERT(w.a != nullptr);
+        GGML_ASSERT(w.b != nullptr);
+        int il = -1;
+        sscanf(cname, "blk.%d.", &il);
+        struct ggml_context * dev_ctx; // device ctx
+        if (il >= 0) {
+            dev_ctx = ctx_map.at(model->buft_layer[il].buft);
+        } else if (strstr(cname, "tok") == 0) {
+            dev_ctx = ctx_map.at(model->buft_input.buft);
+        } else if (strstr(cname, "output") == 0) {
+            dev_ctx = ctx_map.at(model->buft_output.buft);
```

Comment:
You can find the buffer type by looking up the base tensor in `model.tensors_by_name` and calling `ggml_backend_buffer_get_type(t->buffer)`.

Reply:
Fixed on https://github.com/ggerganov/llama.cpp/pull/8332/commits/79e2982788b0102aabb098b1a3d6227a7e32a483